server/api_archive.go

```diff
@@ -0,0 +1,197 @@
+package server
+
+// This is a low-level archiver which zips and sends GCA server persistent file
+// data using a POST command. This is intended to be used by an external service
+// to provide archival snapshots of raw data in the system. Only public data is
+// provided, with private keys zeroed before sending.
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+// The contents of a README file which will be inserted into the archive.
+const ReadmeContents = `This archive contains uninterpreted server files.
+These files all contain publicly available information
+and additional work is needed to stand up a new server
+using them.
+`
+
+// ArchiveHandler provides the POST /archive api endpoint. It returns
+// uninterpreted file data intended to be used by an external service
+// to produce an archive.
+func (gcas *GCAServer) ArchiveHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET method is allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	if r.ContentLength != 0 {
+		http.Error(w, "Request should not have a body", http.StatusBadRequest)
+		return
+	}
+
+	if !gcas.ApiArchiveRateLimiter.Allow() {
+		http.Error(w, fmt.Sprintf("Too many requests, this server allows %v every %v", apiArchiveLimit, apiArchiveRate), http.StatusTooManyRequests)
+		return
+	}
+
+	buffer := new(bytes.Buffer)
+	znw := zip.NewWriter(buffer)
+
+	// File ordering matters here. We are not locking the files, and relying on the file writes being
+	// append only while happening in a single write. We must archive the files in reverse order from
+	// their dependency order.
+
+	for _, name := range PublicFiles {
+		err := gcas.addFile(name, znw)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error zipping file %v: %v", name, err), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Add the pseudo file server.pubkey last, as the other files are signed by it.
+	if err := gcas.addPubKeyFile("server.pubkey", znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error zipping server.pubkey: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	// Add a README
+	if err := gcas.addReadmeFile(znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error adding README: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	if err := znw.Close(); err != nil {
+		if _, err := w.Write(buffer.Bytes()); err != nil {
+			http.Error(w, "Failed to close zip writer", http.StatusInternalServerError)
+			return
+		}
+	}
```

Comment:
We should probably split out all of the parts of the code that build the zip file into its own helper function, I don't like seeing API related code mixed with archiving implementation details.

Reply:
Good suggestion this makes the code easier to understand, and minimizes NEWLINES :smile: